### PR TITLE
fix: resolve pages that double as post type archives

### PR DESCRIPTION
**Pull request by justlevine**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Certain custom post types (e.g. Woocommerce) use a WP `Page` that doubles as a post type archive. Currently querying for these by `page( idType: URI )` returns a `WP_Post_Type` instead of the associated page. 

This PR checks if the provided URI is a post type archive, and if so resolves it to the page, before falling back to returning a `WP_Post_Type`.

Does this close any currently open issues?
------------------------------------------
#1910 , #1850 (I think)

Where has this been tested?
---------------------------
**Operating System:** Ubuntu 20.04 (WSL2 + devilbox)

**WordPress Version:** 5.9

### CODE REVIEW

**Comment by justlevine** [comment]
I dont think those test fails are coming from me...

**Comment by jasonbahl** [comment]
@justlevine I'd love to see some good tests around this to make sure the intended functionality works properly now and to prevent regressions down the road. 

Would you be interested in writing some tests?

**Comment by jasonbahl** [comment]
@justlevine I pulled the code down and tried to run through the WooCommerce "shop" page situation, and either I'm missing a step or this still doesn't address that situation. 

I set up WooCommerce following their wizard, so I have a page named "shop" that can be visited on my WordPress site at `/shop`.

But when I query `nodeByUri( uri: "/shop" ) { __typename }` I get a null response. No node can be found.

**Comment by jasonbahl** [comment]
Possibly the WooCommerce situation wasn't what you optimized for though?

What are the steps to reproduce the other situation? Looking at the issue, it seems as if at least one commenter suggested the issue was resolved: https://github.com/wp-graphql/wp-graphql/issues/1850#issuecomment-880655490

**Comment by justlevine** [comment]
To test:
1. Autogenerate the shop page by installing activating Woocommerce.
2. Disable Woocommerce and rebuild the permalinks.
3. Run the following query:
```graphql
{
  page(id:"/shop", idType:URI) {
    __typename
  }
  nodeByUri(uri:"/shop"){
    __typename
  }
}
```
Both will resolve to page.
4. Enable Woocommerce and expose the `Product` cpt to the schema (add [this gist](https://gist.github.com/justlevine/ffecb991060758a7f5bfdfdfcbbfdf6a) to `mu-plugins`
5. Rerun the query.
`page.__typename` will resolve to `Page`.
`nodeByUri.__typename` will resolve to `ContentType`. (Couldnt think of another way to expose both types for the same identifier, but I'm open to suggestions).

**However**:
6. [🚨]If the `product` post type isnt exposed to WPGraphQL (comment out the code from # 4), both `page` and `nodeByUri` will resolve to `null`. 
7. [⚠] If you set Settings > Reading > Posts page to a static page and then query `nodeByUri(uri:"/my-posts-page")`, `__typename` resolves to `Page`. This is inconsistent with the new behavior.

## tl;dr
Needs some more work and thorough unit testing.

**Comment by jasonbahl** [comment]
@justlevine, your last comment was:

> Needs some more work and thorough unit testing.

Is this still the case, or is this in a state to be reviewed again?

**Comment by justlevine** [comment]
I think anything I tried to do here is a waste of time until we can cleanup/refactor `NodeResolver` (#2342).

**Comment by jasonbahl** [comment]
Ok cool. We'll circle back then
